### PR TITLE
Accelerate data transfer for map Pandas UDF plan

### PR DESCRIPTION
**Pull request by firestarman**
This PR is to accelerate the data transfer between JVM and Python for the plan `GpuMapInPandas`, by implementing the columnar way.

Also add the related integration tests.

For issue #305

### CODE REVIEW

**Comment by firestarman** [comment]
Will run some benchmark locally before marking ready for review.

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
> Will run some benchmark locally before marking ready for review.

Got some positive numbers and tests passed on databricks. Mark as ready for review.

File: integration_tests/src/main/python/udf_test.py
Comment (jlowe) [inline integration_tests/src/main/python/udf_test.py:221]: The return types don't look _that_ limited.  It supports strings, arrays (as long as they don't contain structs or timestamps), structs that don't contain other structs, NullType, DecimalType, etc.  Looks like we're missing a lot of test cases here.

**Review by jlowe — COMMENTED** [review]
I'll defer to @revans2 for reviewing the details of the iterator changes since he's much more familiar with that code, but it does seem like we could do a better job of testing the various type combinations.

File: integration_tests/src/main/python/udf_test.py
Comment (firestarman) [reply integration_tests/src/main/python/udf_test.py:221]: Yes, @revans2  Should we have tests to cover all the output types supported by PySpark ?

**Comment by firestarman** [comment]
> > Got some positive numbers
> 
> Can you elaborate here what those numbers were? What was the use-case and how much improvement did you see?

Not easy to show the numbers here, so here is [the link](https://docs.google.com/spreadsheets/d/1xO_-YalLsl1KIgEdksmBR4UEdmpk5SVNhkFNAuhmLDI/edit#gid=74620754). 

I just ran some benchmark for data transfer locally to verify the columnar way will get better perf,  and this PR is for the issue #305 .

File: integration_tests/src/main/python/udf_test.py
Comment (firestarman) [reply integration_tests/src/main/python/udf_test.py]: I updated the test for the supported output types check.

**Comment by firestarman** [comment]
Forced push only for the commit signature issue. Since the latest merge commit has no signature by default. I added it by a forced push.

**Comment by firestarman** [comment]
build

File: integration_tests/src/main/python/udf_test.py
Comment (jlowe) [inline integration_tests/src/main/python/udf_test.py:221]: That code seems to only indicate that the top-level type is expected to be a struct, not that it doesn't support struct-in-struct?  Poking around at Spark Pandas UDF examples, it looks like struct columns are supported as a result column.

Comment (firestarman) [reply integration_tests/src/main/python/udf_test.py:221]: I verified it locally, and it failed when returning a struct column (even as a top-level column) from Map Pandas UDF.
I uploaded the test code to my repo, you can find it here: https://github.com/firestarman/spark-rapids/commit/02b5b4d4e4a2547fcc35cece04f79d7f34cc70b2 
```bash
        elif self.evalType == PythonEvalType.SQL_MAP_PANDAS_ITER_UDF:
            if isinstance(self._returnType_placeholder, StructType):
                try:
                    to_arrow_type(self._returnType_placeholder)
                except TypeError:
                    raise NotImplementedError(
                        "Invalid return type in mapInPandas: "
>                       "%s is not supported" % str(self._returnType_placeholder))
E                   NotImplementedError: Invalid return type in mapInPandas: StructType(List(StructField(a,StructType(List(StructField(child0,ArrayType(ShortType,true),true),StructField(child1,StructType(List(StructField(child0,ByteType,true),StructField(child1,ShortType,true),StructField(child2,IntegerType,true),StructField(child3,LongType,true),StructField(child4,FloatType,true),StructField(child5,DoubleType,true),StructField(child6,StringType,true),StructField(child7,BooleanType,true),StructField(child8,DateType,true),StructField(child9,TimestampType,true))),true))),true))) is not supported
```

Whether it supports non-nested struct column depends on the type of Pandas UDF, I think. 
e.g. the Scalar Pandas UDF should support top-level struct column, but Map Pandas UDF does not. Since the former one does not require the top-level type being a struct, while the later one does. 
Actually Map Pandas UDF requires returning a DataFrame from UDF, not a column. And the type of a DataFrame is `struct`.

**Review by jlowe — APPROVED** [review]

**Comment by firestarman** [comment]
@revans2 Hi, Could you take a look at this PR ? Thanks in advance.

**Comment by firestarman** [comment]
I am going to merge this, if any concern, I will make new PRs.

**Pull request merged**
